### PR TITLE
Add Asynchronous Metric Instruments SDK

### PR DESCRIPTION
**Pull request by ankit-bhargava**
This PR implements the asynchronous instrument classes described in the Metric Instruments API (PR #161). These classes are capable of actually collecting telemetry data unlike no-op implementations included with the API. Instruments employ Aggregators (pending approval in PR #178 and PR #181 ) to combine the updates they receive into meaningful values. 

A test suite which validates each instrument’s ability to record values in single-threaded and concurrent scenarios is included. Functions necessary for the metrics pipeline as a whole such as reference counting are tested as well.

This PR in conjunction with (PR #179) provides support for all metric instruments listed in the spec. This PR will not compile without all the other PRs mentioned in this description. 

Note: We templated instruments to support a wider array of scalar types. Templating prevents us from using out of line declarations as we normally would. As a result, we left all implementation code in the header file. If there are any suggestions to remedy this we will gladly make the change.  Code for the base classes which asynchronous instruments inherit from is in PR #161.

### CODE REVIEW

File: sdk/include/opentelemetry/sdk/metrics/async_instruments.h [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+       }
+       else
+       {
+         boundAggregators_[labelset]->update(value);
+       }
+       this->mu_.unlock();
+   }
+
+     /*
+      * Activate the intsrument's callback function to record a measurement.  This
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/async_instruments.h:61]:
```suggestion
      * Activate the instrument's callback function to record a measurement.  This
```

File: sdk/include/opentelemetry/sdk/metrics/async_instruments.h [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+#include "opentelemetry/sdk/metrics/instrument.h"
+#include "opentelemetry/metrics/async_instruments.h"
+#include "opentelemetry/version.h"
+#include "opentelemetry/sdk/metrics/aggregator/counter_aggregator.h"
+#include "opentelemetry/sdk/metrics/aggregator/min_max_sum_count_aggregator.h"
+#include <stdexcept>
+#include <map>
+#include <sstream>
+#include <vector>
+#include <memory>
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/async_instruments.h:12]: Minor: would you sort includes alphabetically unless there is a good reason not to do so?

File: sdk/include/opentelemetry/sdk/metrics/async_instruments.h [hunk 3]
```diff
@@ -0,0 +1,225 @@
...
+         for (auto x : boundAggregators_){
+             x.second->checkpoint();
+             ret.push_back(Record(this->GetName(), this->GetDescription(), x.first, x.second));
+         }
+         boundAggregators_.clear();
+         return ret;
+     }
+
+    // Public mapping from labels (stored as strings) to their respective aggregators
+   std::unordered_map<std::string, std::shared_ptr<Aggregator<T>>> boundAggregators_;
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/async_instruments.h:83]: There seems to be indentation issue.

File: sdk/include/opentelemetry/sdk/metrics/async_instruments.h [hunk 4]
```diff
@@ -0,0 +1,225 @@
...
+         if (value < 0){
+             throw std::invalid_argument("Counter instrument updates must be non-negative.");
+         } else {
+             sp1->update(value);
+         }
+       }
+       else
+       {
+           if (value < 0){
+               throw std::invalid_argument("Counter instrument updates must be non-negative.");
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/async_instruments.h:125]: @pyohannes do you think we should check `__EXCEPTIONS` and call `std::terminate`?
I think we shouldn't throw if exception is disabled, however we don't want to kill the process as this is not a critical failure.

File: sdk/include/opentelemetry/sdk/metrics/async_instruments.h [hunk 5]
```diff
@@ -0,0 +1,225 @@
...
+         std::vector<Record> ret;
+         for (auto x : boundAggregators_){
+             x.second->checkpoint();
+             ret.push_back(Record(this->GetName(), this->GetDescription(), x.first, x.second));
+         }
+         boundAggregators_.clear()
+         return ret;
+     }
+
+    // Public mapping from labels (stored as strings) to their respective aggregators
```
Comment (reyang) [inline sdk/include/opentelemetry/sdk/metrics/async_instruments.h:219]: This seems to be another indentation issue, would you rebase and check if the formatter CI would pass?

File: sdk/include/opentelemetry/sdk/metrics/async_instruments.h (hunk 4 above)
Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/async_instruments.h]: I've added a check and call to terminate :)

**Comment by ankit-bhargava** [comment]
@reyang Thanks for the review! I rebased and addressed your comments.  Additionally, we discovered a bug related to the way we were dealing with concurrency in our synchronous instruments.  I added fixes in this PR out of convenience (you can see the changes in the `sync_instruments.h` file, but I can remove them and open a new PR if you would prefer that.

File: sdk/include/opentelemetry/sdk/metrics/async_instruments.h (hunk 2 above)
Comment (ankit-bhargava) [reply sdk/include/opentelemetry/sdk/metrics/async_instruments.h]: The format tools reorder the includes but I sorted them where I could.

**Comment by ankit-bhargava** [comment]
@reyang This should be ready to merge unless there are any further concerns

**Pull request merged**